atintegrators/ExactHamiltonianPass.cc

```diff
@@ -0,0 +1,207 @@
+                     int num_particles)
+{
+    int c, n;
+    double * r6;
+
+    // copy AT structure into pass method structure
+
+    element e = {0};
+
+    for(n = 0; n < max_order; n++)
+    {
+        e.F[2 * n]     = B[n];
+        e.F[2 * n + 1] = A[n];
+    }
+    
+    e.L = le;
+    e.phi = phi;
+    e.gK = gK;
+    e.nF = max_order;
+    e.slices = num_int_steps;
+    e.type = type;
+    e.do_multipole_fringe = multipole_fringe;
+
+    for(c = 0; c<num_particles; c++)
+    {
+        r6 = r_in+c*6;	
+        if(!atIsNaN(r6[0]))
+        {   
+            /* misalignment at entrance */
+            if(T1)
+                ATaddvv(r6,T1);
+            if(R1)
+                ATmultmv(r6,R1);
+            
+            track_element(r6, &e);
+            
+            /* misalignment at exit */
+            if(R2)
+                ATmultmv(r6,R2);
+            if(T2)   
+                ATaddvv(r6,T2);
+        }
+    }
+}
+
+#if defined(MATLAB_MEX_FILE) || defined(PYAT)
+
+ExportMode struct elem *trackFunction(const atElem *ElemData, struct elem *Elem,
+                               double *r_in, int num_particles,
+                               struct parameters *Param)
+
+{
+    double le, bending_angle;
+    double *polynom_a, *polynom_b;
+    long max_order, num_int_steps, type, multipole_fringe;
+    double *R1, *R2, *T1, *T2;
+    double phi, gK;
+    le = atGetDouble(ElemData,"Length"); check_error();
+    bending_angle = atGetDouble(ElemData,"BendingAngle"); check_error();
+    polynom_a = atGetDoubleArray(ElemData,"PolynomA"); check_error();
+    polynom_b = atGetDoubleArray(ElemData,"PolynomB"); check_error();
+    max_order = atGetLong(ElemData, "MaxOrder"); check_error();
+    num_int_steps = atGetLong(ElemData, "NumIntSteps"); check_error();
+    type = atGetLong(ElemData, "Type"); check_error();
+    multipole_fringe = atGetLong(ElemData, "Type"); check_error();
+    phi = atGetDouble(ElemData,"Phi"); check_error();
+    gK = atGetDouble(ElemData,"gK"); check_error();
+    R1=atGetOptionalDoubleArray(ElemData,"R1"); check_error();
+    R2=atGetOptionalDoubleArray(ElemData,"R2"); check_error();
+    T1=atGetOptionalDoubleArray(ElemData,"T1"); check_error();
+    T2=atGetOptionalDoubleArray(ElemData,"T2"); check_error();
+    ExactHamiltonianPass(r_in, le, polynom_a, polynom_b, T1, T2, R1, R2, max_order,
+             num_int_steps, phi, type, gK, multipole_fringe, num_particles);
+    Elem = (struct elem*)atMalloc(sizeof(struct elem));
+    Elem->Length=le;
+    Elem->BendingAngle=bending_angle;
+    Elem->PolynomA=polynom_a;
+    Elem->PolynomB=polynom_b;
+    Elem->MaxOrder=max_order;
+    Elem->NumIntSteps=num_int_steps;
+    /*optional fields*/
+    Elem->R1=R1;
+    Elem->R2=R2;
+    Elem->T1=T1;
+    Elem->T2=T2;
+
+    return Elem;
+}
+
+MODULE_DEF(ExactHamiltonianPass) /* Dummy module initialisation */
+
+#endif /* defined(PYAT) || defined(MATLAB_MEX_FILE) */
+
+
+
+#if defined(MATLAB_MEX_FILE)
```

Comment:
The following does not compile on Matlab. This whole Matlab-specific part should be replaced by:

```
#if defined(MATLAB_MEX_FILE)
void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
{
    if (nrhs >= 2) {
        double * r_in;
        const mxArray * ElemData = prhs[0];
        int num_particles = mxGetN(prhs[1]);
        int m = mxGetM(prhs[1]);
        if (m!=6)
            mexErrMsgTxt("Second argument must be a 6 x N matrix");

        double le, bending_angle;
        double *polynom_a, *polynom_b;
        long max_order, num_int_steps, type, multipole_fringe;
        double *R1, *R2, *T1, *T2;
        double phi, gK;
        le = atGetDouble(ElemData,"Length"); check_error();
        bending_angle = atGetDouble(ElemData,"BendingAngle"); check_error();
        polynom_a = atGetDoubleArray(ElemData,"PolynomA"); check_error();
        polynom_b = atGetDoubleArray(ElemData,"PolynomB"); check_error();
        max_order = atGetLong(ElemData, "MaxOrder"); check_error();
        num_int_steps = atGetLong(ElemData, "NumIntSteps"); check_error();
        type = atGetLong(ElemData, "Type"); check_error();
        multipole_fringe = atGetLong(ElemData, "Type"); check_error();
        phi = atGetDouble(ElemData,"Phi"); check_error();
        gK = atGetDouble(ElemData,"gK"); check_error();
        /*optional fields*/
        R1=atGetOptionalDoubleArray(ElemData,"R1"); check_error();
        R2=atGetOptionalDoubleArray(ElemData,"R2"); check_error();
        T1=atGetOptionalDoubleArray(ElemData,"T1"); check_error();
        T2=atGetOptionalDoubleArray(ElemData,"T2"); check_error();

        /* ALLOCATE memory for the output array of the same size as the input  */
        plhs[0] = mxDuplicateArray(prhs[1]);
        r_in = mxGetPr(plhs[0]);
        ExactHamiltonianPass(r_in, le, polynom_a, polynom_b, T1, T2, R1, R2, max_order,
                     num_int_steps, phi, type, gK, multipole_fringe, num_particles);
    }
    else if (nrhs == 0) {
        /* list of required fields */
        plhs[0] = mxCreateCellMatrix(5,1);
        mxSetCell(plhs[0],0,mxCreateString("Length"));
        mxSetCell(plhs[0],0,mxCreateString("BendingAngle"));
        mxSetCell(plhs[0],1,mxCreateString("PolynomA"));
        mxSetCell(plhs[0],2,mxCreateString("PolynomB"));
        mxSetCell(plhs[0],3,mxCreateString("MaxOrder"));
        mxSetCell(plhs[0],4,mxCreateString("NumIntSteps"));
        mxSetCell(plhs[0],4,mxCreateString("Type"));
        mxSetCell(plhs[0],4,mxCreateString("Type"));
        mxSetCell(plhs[0],4,mxCreateString("Phi"));
        mxSetCell(plhs[0],4,mxCreateString("gK"));
        if (nlhs>1) {
            /* list of optional fields */
            plhs[1] = mxCreateCellMatrix(11,1);
            mxSetCell(plhs[1], 4,mxCreateString("T1"));
            mxSetCell(plhs[1], 5,mxCreateString("T2"));
            mxSetCell(plhs[1], 6,mxCreateString("R1"));
            mxSetCell(plhs[1], 7,mxCreateString("R2"));
        }
    }
    else {
        mexErrMsgIdAndTxt("AT:WrongArg","Needs 0 or 2 arguments");
    }
}
#endif /*MATLAB_MEX_FILE*/

```